internal/grader/regrade.go

```diff
@@ -0,0 +1,229 @@
+	RegradeAfter *timestamp.Timestamp `json:"regrade-after"`
+
+	// If true, do not swap the context to the background context when running.
+	// By default (when this is false), the context will be swapped to the background context when !WaitForCompletion.
+	// The swap is so that regrade does not get canceled when an HTTP request is complete.
+	// Setting this true is useful for testing (as one round of analysis tests can be wrapped up).
+	RetainOriginalContext bool `json:"-"`
+
+	ResolvedUsers []string `json:"-"`
+}
+
+type RegradeResult struct {
+	Options      RegradeOptions                          `json:"options"`
+	RegradeAfter timestamp.Timestamp                     `json:"regrade-after"`
+	Results      map[string]*model.SubmissionHistoryItem `json:"results"`
+	WorkErrors   map[string]string                       `json:"work-errors"`
+}
+
+func Regrade(assignment *model.Assignment, options RegradeOptions) (*RegradeResult, int, error) {
+	reference, err := model.ParseCourseUserReferences(options.RawReferences)
+	if err != nil {
+		return nil, 0, fmt.Errorf("Failed to parse course user references: '%w'.", err)
+	}
+
+	courseUsers, err := db.GetCourseUsers(assignment.GetCourse())
+	if err != nil {
+		return nil, 0, fmt.Errorf("Failed to get course users: '%w'.", err)
+	}
+
+	options.ResolvedUsers = model.ResolveCourseUserEmails(courseUsers, reference)
+
+	if !options.RetainOriginalContext && !options.WaitForCompletion {
+		options.Context = context.Background()
+	}
+
+	if options.RegradeAfter == nil {
+		now := timestamp.Now()
+		options.RegradeAfter = &now
+	}
+
+	// Lock based on the course to prevent multiple requests using up all the cores.
+	lockKey := fmt.Sprintf("regrade-course-%s", assignment.GetCourse().GetID())
+
+	job := jobmanager.Job[string, *model.SubmissionHistoryItem]{
+		JobOptions:              &options.JobOptions,
+		LockKey:                 lockKey,
+		PoolSize:                config.REGRADE_COURSE_POOL_SIZE.Get(),
+		ReturnIncompleteResults: !options.WaitForCompletion,
+		WorkItems:               options.ResolvedUsers,
+		RetrieveFunc: func(resolvedEmails []string) (map[string]*model.SubmissionHistoryItem, error) {
+			return retrieveRegradedSubmissions(assignment, *options.RegradeAfter, resolvedEmails)
+		},
+		WorkFunc: func(email string) (*model.SubmissionHistoryItem, error) {
+			return performSingleRegrade(courseUsers, assignment, options, email)
+		},
+		WorkItemKeyFunc: func(email string) string {
+			return fmt.Sprintf("%s-%s", lockKey, email)
+		},
+	}
+
+	err = job.Validate()
+	if err != nil {
+		return nil, 0, fmt.Errorf("Failed to validate job: '%w'.", err)
+	}
+
+	output := job.Run()
+	if output.Error != nil {
+		return nil, 0, fmt.Errorf("Failed to run regrade job '%s': '%w'.", output.ID, output.Error)
+	}
+
+	workErrors := make(map[string]string, len(output.WorkErrors))
+	for email, err := range output.WorkErrors {
+		workErrors[email] = err.Error()
+
+		logAttributes := make([]any, 3)
```

Comment:
This allocation seems like it could be off to me.
 - Why 3?
 - The length is being set, when it seems like it should be the capacity.